server/svix-server/src/cfg.rs

```diff
@@ -206,37 +214,36 @@ impl ConfigurationInner {
     /// Fetches the configured backend information for the queue. May panic is the configuration has
     /// not been validated
     pub fn queue_backend(&self) -> QueueBackend<'_> {
-        match self.queue_type {
-            QueueType::Memory => QueueBackend::Memory,
-            QueueType::Redis => QueueBackend::Redis(
+        let dsn = || {
+            self.queue_dsn.as_ref().unwrap_or_else(|| {
                 self.redis_dsn
                     .as_ref()
-                    .expect("Called [`queue_backend`] before validating configuration"),
-            ),
-            QueueType::RedisCluster => QueueBackend::RedisCluster(
-                self.redis_dsn
-                    .as_ref()
-                    .expect("Called [`queue_backend`] before validating configuration"),
-            ),
+                    .expect("Called [`queue_backend`] before validating configuration")
+            })
+        };
+
+        match self.queue_type {
+            QueueType::Memory => QueueBackend::Memory,
+            QueueType::Redis => QueueBackend::Redis(dsn()),
+            QueueType::RedisCluster => QueueBackend::RedisCluster(dsn()),
         }
     }
 
     /// Fetches the configured backend information for the cache, or `None` if the [`CacheType`] is
     ///  `None`. May panic is the configuration has not been validated
     pub fn cache_backend(&self) -> CacheBackend<'_> {
+        let dsn = || {
```

Comment:
Why don't just make it a struct function and then also use it in the validation function?